augur/filter/_run.py

```diff
@@ -1,502 +1,45 @@
-        if args.output:
-            # Get the samples to be deleted, not to keep, for VCF
-            dropped_samps = list(sequence_strains - valid_strains)
-            write_vcf(args.sequences, args.output, dropped_samps)
-    elif args.sequences:
-        sequences = read_sequences(args.sequences)
-
-        # If the user requested sequence output, stream to disk all sequences
-        # that passed all filters to avoid reading sequences into memory first.
-        # Even if we aren't emitting sequences, we track the observed strain
-        # names in the sequence file as part of the single pass to allow
-        # comparison with the provided sequence index.
-        if args.output:
-            observed_sequence_strains = set()
-            with open_file(args.output, "wt") as output_handle:
-                for sequence in sequences:
-                    observed_sequence_strains.add(sequence.id)
-
-                    if sequence.id in valid_strains:
-                        write_sequences(sequence, output_handle, 'fasta')
-        else:
-            observed_sequence_strains = {sequence.id for sequence in sequences}
+from .report import print_report
+from .subsample import apply_subsampling
 
-        if sequence_strains != observed_sequence_strains:
-            # Warn the user if the expected strains from the sequence index are
-            # not a superset of the observed strains.
-            if sequence_strains is not None and observed_sequence_strains > sequence_strains:
-                print_err(
-                    "WARNING: The sequence index is out of sync with the provided sequences.",
-                    "Metadata and strain output may not match sequence output."
-                )
 
-            # Update the set of available sequence strains.
-            sequence_strains = observed_sequence_strains
+def run(args: Namespace):
+    with NamedTemporaryFile() as file:
+        # Set the database file as a variable that can be easily accessed within
+        # functions deep in the call stack. It could be passed down by value,
+        # but that would be tedious and makes it harder to trace references back
+        # to the source.
+        constants.RUNTIME_DB_FILE = file.name
 
-    # Calculate the number of strains that don't exist in either metadata or
-    # sequences.
-    num_excluded_by_lack_of_metadata = 0
-    if sequence_strains:
-        # Update strains to keep based on available sequence data. This prevents
-        # writing out strain lists or metadata for strains that have no
-        # sequences.
-        valid_strains = valid_strains & sequence_strains
+        if args.debug:
+            constants.RUNTIME_DEBUG = True
 
-        num_excluded_by_lack_of_metadata = len(sequence_strains - metadata_strains)
+        initialize_input_source_table()
 
-    if args.output_strains:
-        output_strains.close()
+        import_metadata(args.metadata, args.metadata_id_columns, args.metadata_delimiters)
```

Comment:
One significant change in this rewrite is that the metadata is no longer processed in chunks. This has implementation benefits, but comes with the downside of substantially increased disk usage.

Previously, there was the issue of `augur filter` using too much memory, which prompted the current chunked implementation. Now, I'm worried about `augur filter` using too much disk space, especially for large datasets such as [ncov/open/metadata.tsv.zst](https://data.nextstrain.org/files/ncov/open/metadata.tsv.zst), which results in a 12GB database file based on a trial run just now. This should be addressed before merging.

Two ideas to address this:

1. Make improvements to minimize the amount of size of the database file. The easy win here is to read a subset of the metadata columns, since most usages of `augur filter` only use a few columns for computation, while the remainder is only used for final output. I started on this in bc2a0305d0f5e09c1b2955f852a424489689f0e3. The difficult part will be determining which columns are used by `--query-pandas`/`--query-sqlite`.
2. Warn users if they don't have enough disk space. This requires an estimation of how large the database file will be, which isn't trivial. I started on this in b411c589644e1ffaa18eec00da449238bf2ca1c7.

I think (1) is worthwhile, and might remove the need for (2) if disk usage is reduced significantly.

Reply:
Generally concur with you here. Some ideas for implementing the disk usage reduction mitigations of (1)…

For determining which columns are used by `--query-sqlite`, two possibilities come to mind:

1. Parse the SQL query into tokens/an AST and inspect that. Two examples of where I did this previously using the `sqlparse` Python package: [checking usage of transactions](https://github.com/seattleflu/id3c/blob/76a7f7e68f86e893fefdaf8deb047dd4407a63b4/tests/sqitch-transactions.py) and [checking usage of `SET search_path` statements](https://github.com/seattleflu/id3c/blob/76a7f7e68f86e893fefdaf8deb047dd4407a63b4/tests/search-path.py). While both split a SQL script into a list of statements, they take two tacks from there: the first walks the tokens in each statement, the second matches a regex against each statement. We'd probably want the first tack here.
2. Use SQLite's introspection capabilities to ask what the query will use/do (e.g. with `EXPLAIN` or other methods). I've done this manually before in other database systems (with structured `EXPLAIN` output), but not programmatically in SQLite. It may not be possible and/or feasible, but seems worth looking into esp. if (1) isn't workable for some reason.

For determining what's used by `--query-pandas`, I'd think to either:

1. Leverage Pandas' internals here (if reasonably possible), as I believe you've prototyped elsewhere…
2. Implement a minimal parser for the Pandas query grammar, e.g. using pyparsing. This doesn't have to support the whole grammar, just what's sufficient to extract column names. In a similar vein, I took this approach for [Markdown modification in Nextstrain CLI](https://github.com/nextstrain/cli/blob/630ba6fd880b8fcc8677ea96c9375621856ed54f/nextstrain/cli/markdown.py#L51-L192).
3. Rig up a `pandas.DataFrame` with monkeypatching, subclassing, or a proxy object that monitors and inspects column access and then apply the query to it.

Reply:
@tsibley thanks for the suggestions! As you [mentioned elsewhere](https://github.com/nextstrain/augur/pull/1278#issuecomment-1690696870), Python's `ast` library seems to work well for pandas queries. I've successfully implemented (1) loading a subset of columns in #1294. If that gets merged, I'll rework this PR around those changes.

Reply:
Updates:

- For `--query-sqlite`, I've started implementation using `sqlparse` in 3cf5670d8c4207c8ccce44b6b5f10f66d5eaac76.
- For `--query-pandas` using `ast`, that was recently added in 0c5a13724f17a267009cb18a0bd834f50ab90508, released, broke some workflows, and now we are trying to leverage Pandas internals ([again](https://github.com/nextstrain/augur/commit/2ead5b3e3306dc1100b49eb774287496018122d9), but hopefully [better](https://github.com/nextstrain/augur/pull/1416)).